Open a specific page in the layer properties dialog using the iface instance

suricactus opened the pull request:
Everyone can enjoy a call like this:
```
from qgis.core import Qgis
from qgis.utils import iface

if Qgis.QGIS_VERSION_INT >= 32000:
    iface.showLayerProperties(iface.activeLayer(), 'mOptsPage_Style')
else:
    iface.showLayerProperties(iface.activeLayer())
```

suricactus committed:
Open a specific page in the layer properties dialog using the iface instance

roya0045 commented:
Are the acceptable inputs documented, or at least the pattern? 'mOptsPage_Style' is not something that is intuitive, if that's even a thing.

suricactus commented:
AFAIK the default settings pages are not documented, but when you add a custom layer properties page you know it in advance.

m-kuhn commented:
These are the currently valid pages, I'd just document these, providing a future proof solution is hard:

```
# Mesh Layer
mOptsPage_Information
mOptsPage_Source
mOptsPage_Style
mOptsPage_StyleContent
mOptsPage_Rendering
mOptsPage_Temporal
mOptsPage_Metadata

# Point Cloud Layer
mOptsPage_Information
mOptsPage_Source
mOptsPage_Metadata
mOptsPage_Statistics

# Raster Layer
mOptsPage_Information
mOptsPage_Source
mOptsPage_Style
mOptsPage_Transparency
mOptsPage_Histogram
mOptsPage_Rendering
mOptsPage_Temporal
mOptsPage_Pyramids
mOptsPage_Metadata
mOptsPage_Legend
mOptsPage_Server
mOptsPage_Information
mOptsPage_Source
mOptsPage_Style
mOptsPage_Labels
mOptsPage_Masks
mOptsPage_Diagrams
mOptsPage_SourceFields
mOptsPage_AttributesForm
mOptsPage_Joins
mOptsPage_AuxiliaryStorage
mOptsPage_Actions
mOptsPage_Display
mOptsPage_Rendering
mOptsPage_Temporal
mOptsPage_Variables
mOptsPage_Metadata
mOptsPage_DataDependencies
mOptsPage_Legend
mOptsPage_Server

# Vector Tile Layer
mOptsPage_Information
mOptsPage_Style
mOptsPage_Labeling
mOptsPage_Metadata

```

m-kuhn commented on src/gui/qgisinterface.h:
```diff
@@ -972,7 +972,7 @@ class GUI_EXPORT QgisInterface : public QObject
     virtual void removeDockWidget( QDockWidget *dockwidget ) = 0;
 
     //! Open layer properties dialog
```
```suggestion
    /**
     * Opens layer properties dialog for the layer provided with \a l
     * A \a page to open can be specified (supported since QGIS 3.20).
     * Valid strings are [...], others may be available depending on the context.
     */
```

nyalldawson replied:
I'd also note that the strings are subject to change without notice between QGIS versions, and that the tab names themselves aren't considered part of the stable API.

suricactus replied:
Should be better now.

uclaros commented:
We could organize those in an enum to handle the pages.

roya0045 commented:
> We could organize those in an enum to handle the pages.

I think the optimal option would be to have a map with translated keys, this way the used can go with the terms they are most common with to get the layer and convert it to lowercase to prevent case mismatch shenanigans when  looking up.

m-kuhn commented:
> We could organize those in an enum to handle the pages.

It's possible to add custom pages with plugins, so there's also a limitation to that.

uclaros commented:
> It's possible to add custom pages with plugins, so there's also a limitation to that.

Right, I thought about the QFieldSync tab right after posting...

suricactus committed:
Added docstring describing the page argument of showLayerProperties

One can get a list of page names by executing the following command:
grep \"mOptsPage ./build/src/ui/ui_qgs*layerpropertiesbase.h

m-kuhn commented on python/gui/auto_generated/qgisinterface.sip.in:
```diff
@@ -1138,9 +1138,42 @@ Removes the specified ``dock`` widget from main window (without deleting it).
+.. note::
+
+   Page names are subject to change without notice between QGIS versions,
+   and that the tab names themselves aren't considered part of the stable API.
```
```suggestion
   they are not considered part of the stable API.
```

m-kuhn commented on src/gui/qgisinterface.h:
```diff
@@ -971,8 +971,39 @@ class GUI_EXPORT QgisInterface : public QObject
+     * mOptsPage_Information, mOptsPage_Style, mOptsPage_Labeling, mOptsPage_Metadata
+     *
+     * \note Page names are subject to change without notice between QGIS versions,
+     * and that the tab names themselves aren't considered part of the stable API.
```
```suggestion
     * they are not considered part of the stable API.
```

suricactus committed:
Update src/gui/qgisinterface.h

Co-authored-by: Matthias Kuhn <matthias@opengis.ch>

suricactus committed:
Update python/gui/auto_generated/qgisinterface.sip.in

Co-authored-by: Matthias Kuhn <matthias@opengis.ch>
